bootloader/grub.go

```diff
@@ -379,11 +385,97 @@ func (g *grub) ManagedAssets() []string {
+		isRecovery = true
+	}
+	staticCmdline, err := staticCommandLineFromGrubAsset(assetName)
+	if err != nil {
+		return "", fmt.Errorf("cannot extract static command line element: %v", err)
+	}
+	// split the mode arguments and make sure they are sane for given
+	// run/recovery mode
+	mode, err := modeArgsForGrub(modeArgs, isRecovery)
+	if err != nil {
+		return "", err
+	}
+	args, err := strutil.KernelCommandLineSplit(staticCmdline + " " + extraArgs)
+	if err != nil {
+		return "", fmt.Errorf("cannot use badly formatted kernel command line: %v", err)
+	}
+	// join all argument with a single space, see
+	// grub-core/lib/cmdline.c:grub_create_loader_cmdline() for reference,
+	// arguments are separated by a single space, the space after last is
+	// replaced with terminating NULL
+	return strings.Join(append(mode, args...), " "), nil
+}
+
+// static command line is defined as:
+//     set snapd_static_cmdline_args='arg arg arg'\n
+// or
+//     set snapd_static_cmdline_args='arg'\n
+// or (empty static command line)
+//     set snapd_static_cmdline_args=''\n
+//     set snapd_static_cmdline_args=\n
+var grubStaticCmdlineArgsRE = regexp.MustCompile(`(?m)^set snapd_static_cmdline_args=(.*)$`)
+
+// staticCommandLineFromGrubAsset extracts the static command line element from
+// grub boot config asset on disk.
+func staticCommandLineFromGrubAsset(asset string) (string, error) {
+	gbc := assets.Internal(asset)
+	if gbc == nil {
+		return "", fmt.Errorf("internal error: asset %q not found", asset)
+	}
+	m := grubStaticCmdlineArgsRE.FindSubmatch(gbc)
+	if len(m) == 2 {
+		cmdline := string(m[1])
+		if cmdline == "" || cmdline == "''" {
+			return "", nil
+		}
+		if len(cmdline) < 2 || cmdline[0] != '\'' || cmdline[len(cmdline)-1] != '\'' {
+			return "", fmt.Errorf("incorrect static command line format: %q", m[0])
+		}
+		return cmdline[1 : len(cmdline)-1], nil
+	}
+	return "", nil
+}
+
+// modeArgsForGrub validates the snapd mode arguments for the kernel command
+// line and splits them into individual arguments.
+func modeArgsForGrub(modeArgs string, recoveryAsset bool) ([]string, error) {
+	args, err := strutil.KernelCommandLineSplit(modeArgs)
+	if err != nil {
+		return nil, err
+	}
+	// see grub.cfg and grub-recovery.cfg assets, the order is:
+	//      for run mode: snapd_recovery_mode=run <args>
+	// for recovery mode: snapd_recovery_mode=recover snapd_recovery_system=<label> <args>
+	if !recoveryAsset {
+		if len(args) != 1 || args[0] != "snapd_recovery_mode=run" {
+			return nil, fmt.Errorf("unexpected run asset mode arguments: %q", modeArgs)
+		}
+		return args[:1], nil
+	} else {
+		if len(args) != 2 {
+			return nil, fmt.Errorf("unexpected recovery asset mode arguments: %q", modeArgs)
+		}
+		recover, system := args[0], args[1]
+		if args[0] != "snapd_recovery_mode=recover" {
+			system, recover = recover, system
+		}
+		if recover != "snapd_recovery_mode=recover" || !strings.HasPrefix(system, "snapd_recovery_system=") {
```

Comment:
Nitpick, could use a helper `var recovery = "snapd_recovery_mode=recover"` to avoid repetition in these two ifs.